source/base/partitioner.cc

```diff
@@ -209,6 +211,15 @@ namespace Utilities
                                     communicator);
         AssertThrowMPI(ierr);
       }
+
+      // Allow non-zero start index for the vector. Part 2:
+      // We correct the assumption made above and let the
+      // index set of rank 0 actually start from the
+      // correct value, i.e. we correct the shift to
+      // its start.
+      if (my_pid == 0)
+        my_shift = local_range_data.first;
+
       if (my_shift != local_range_data.first)
```

Comment:
When we have gaps *inside* the index set, I believe this code is still wrong. The reason why we have this block is to fix the local range for the case we do not have any indices in the local range and hence reading a range from the associated index set is not possible. Otherwise, `local_range_data` should already be correct and the data desired by the user. To fix this, I suggest we write:
```
// Fix the index start in case the index set could not give us that information.
if (local_range_data.first == 0 && my_shift != 0)
```

Reply:
Looks good. Although, I am not really sure why support this gap (only) at the beginning of the vector...

Reply:
The problem is that we make the assumption elsewhere in this class that the local range is contiguous. That could be relaxed, but it is not necessary in this PR.

Reply:
1. Is there an Assert at the top for making sure the local range is contiguous? If not, we should add that.
2. @kronbichler , I think @peterrum is referring to the situation of having a gap between processors (with each processor having a contiguous range).

Reply:
3. I am still not sure I understand this block of code. Is it for gaps between processors, the gap before proc 0, or for processors with 0 owned indices? The comment doesn't explain this.